columnflow/tasks/framework/base.py

```diff
@@ -1434,96 +1459,95 @@ def resolve_shifts(cls, params: dict[str, Any]) -> dict:
+            if (
+                params["shift"] not in {law.NO_STR, None} and
+                params["local_shift"] not in {law.NO_STR, None} and
+                (not params.get("global_shift_insts") or not params.get("local_shift_insts"))
+            ):
+                params["global_shift_insts"] = {}
+                params["local_shift_insts"] = {}
+                for config_inst, (gs, ls) in shift_pairs.items():
+                    params["global_shift_insts"][config_inst] = config_inst.get_shift(gs)
+                    params["local_shift_insts"][config_inst] = config_inst.get_shift(ls)
 
         return params
 
     @classmethod
-    def _resolve_shifts(cls, config_inst: od.Config, params: dict) -> tuple[str, str] | None:
-        # require that the shift is set and known
-        if (requested_shift := params.get("shift")) in (None, law.NO_STR):
-            if not cls.allow_empty_shift:
-                raise Exception(f"no shift found in params: {params}")
-            return (law.NO_STR, law.NO_STR)
-
-        if requested_shift not in config_inst.shifts:
-            return None
-
-        # actual shift resolution: compare the requested shift to known ones
-        # local_shift -> the requested shift if implemented by the task itself, else nominal
-        # shift       -> the requested shift if implemented by this task
-        #                or an upsteam task (== global shift), else nominal
-        global_shift = requested_shift
-        if (local_shift := params.get("local_shift")) in {None, law.NO_STR}:
-            # determine the known shifts for this class
-            shifts = TaskShifts()
-            cls.get_known_shifts(params, shifts)
-            # check cases
-            if requested_shift in shifts.local:
-                local_shift = requested_shift
-            elif requested_shift in shifts.upstream:
-                local_shift = "nominal"
-            else:
-                global_shift = "nominal"
-                local_shift = "nominal"
-
-        return global_shift, local_shift
+    def _resolve_upstream_shifts(cls, config_inst: od.Config, params: dict[str, Any]) -> tuple[str | None, str | None]:
+        # the shift resolution task must be set and support a single config
+        if cls.shift_resolution_task_cls is None:
+            raise AttributeError(
+                f"shift resolution of multi-config task {cls} requires 'shift_resolution_task_cls' to be set",
+            )
+        if not cls.shift_resolution_task_cls.has_single_config():
+            raise ValueError(
+                f"shift_resolution_task_cls {cls.shift_resolution_task_cls} must support a single config",
+            )
+        # run the parameter resolution and return obtained shifts
+        _params = params | {
+            "config": config_inst.name,
+            "config_inst": config_inst,
+        }
+        cls.shift_resolution_task_cls.resolve_shifts(_params, allow_missing=True)
```

Comment:
1.) the `cls.shift_resolution_task_cls.resolve_shifts` will never be called when `cls` is a single config task
2.) at which stage do we call the `cls.shift_resolution_task_cls.resolve_shifts`per dataset? To me it seems that this is only being done for the `DatasetsProcessesShiftSourcesMixin`. Do we also need a `DatasetsProcessesShiftMixin` doing something similar?

Reply:
1) I'm going to add this right now :+1: Regarding the similarity between the sc and mc handling, I wanted to keep them separate for now, and streamline it after we managed to make it behave the way we want.
2) Indeed, this is to be done by subclasses just like we do with shift sources. `DatasetsProcessesShiftMixin` sounds reasonable and would fall right into place. I can create a shallow task in *this* PR, but I would leave it to the next one to use it in the single shift plot base.